File: cocos/physics/physx/physx-rigid-body.ts
/*
 Copyright (c) 2020 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

/**
 * @packageDocumentation
 * @hidden
 */

/* eslint-disable @typescript-eslint/no-unsafe-return */
import { IVec3Like, Vec3 } from '../../core';
import { ERigidBodyType, PhysicsSystem, RigidBody } from '../framework';
import { IRigidBody } from '../spec/i-rigid-body';
import { applyForce, applyImpulse, applyTorqueForce, PX, _trans } from './export-physx';
import { PhysXSharedBody } from './physx-shared-body';
import { PhysXWorld } from './physx-world';

const v3_0 = new Vec3();

export class PhysXRigidBody implements IRigidBody {
    get impl (): any { return this._sharedBody.impl; }

    get isAwake (): boolean { return !this.isStatic && !this.impl.isSleeping(); }
    get isSleeping (): boolean { return this.isStatic || this.impl.isSleeping(); }

    get isEnabled (): boolean { return this._isEnabled; }
    get rigidBody (): RigidBody { return this._rigidBody; }
    get sharedBody (): PhysXSharedBody { return this._sharedBody; }
    get isStatic (): boolean { return !this.impl || this._sharedBody.isStatic; }
    get isStaticOrKinematic (): boolean { return !this.impl || this._sharedBody.isStatic || this._sharedBody.isKinematic; }
    get isInScene (): boolean { return this._sharedBody.isInScene; }

    isSleepy = false;
    private _isEnabled = false;
    private _rigidBody!: RigidBody;
    private _sharedBody!: PhysXSharedBody;

    initialize (v: RigidBody): void {
        this._rigidBody = v;
        this._sharedBody = (PhysicsSystem.instance.physicsWorld as PhysXWorld).getSharedBody(v.node, this);
        this._sharedBody.reference = true;
    }

    onEnable (): void {
        this._isEnabled = true;
        this.setMass(this._rigidBody.mass);
        this.setType(this._rigidBody.type);
        this.setAllowSleep(this._rigidBody.allowSleep);
        this.setLinearDamping(this._rigidBody.linearDamping);
        this.setAngularDamping(this._rigidBody.angularDamping);
        this.setLinearFactor(this._rigidBody.linearFactor);
        this.setAngularFactor(this._rigidBody.angularFactor);
        this.useGravity(this._rigidBody.useGravity);
        this._sharedBody.enabled = true;
    }

    onDisable (): void {
        this._isEnabled = false;
        this._sharedBody.enabled = false;
    }

    onDestroy (): void {
        this._sharedBody.reference = false;
        (this._rigidBody as any) = null;
        (this._sharedBody as any) = null;
    }

    setType (v: ERigidBodyType): void {
        this._sharedBody.setType(v);
    }

    setMass (v: number): void {
        if (this.isStatic) return;
        this._sharedBody.setMass(v);
    }

    setLinearDamping (v: number): void {
        this._sharedBody.setLinearDamping(v);
    }

    setAngularDamping (v: number): void {
        this._sharedBody.setAngularDamping(v);
    }

    useGravity (v: boolean): void {
        if (this.isStatic) return;
        this.impl.setActorFlag(PX.ActorFlag.eDISABLE_GRAVITY, !v);
    }

    useCCD (v: boolean): void {
        if (this.isStatic) return;
        this.impl.setRigidBodyFlag(PX.RigidBodyFlag.eENABLE_CCD, v);
    }

    setLinearFactor (v: IVec3Like): void {
        if (this.isStatic) return;
        this.impl.setRigidDynamicLockFlag(PX.RigidDynamicLockFlag.eLOCK_LINEAR_X, !v.x);
        this.impl.setRigidDynamicLockFlag(PX.RigidDynamicLockFlag.eLOCK_LINEAR_Y, !v.y);
        this.impl.setRigidDynamicLockFlag(PX.RigidDynamicLockFlag.eLOCK_LINEAR_Z, !v.z);
    }

    setAngularFactor (v: IVec3Like): void {
        if (this.isStatic) return;
        this.impl.setRigidDynamicLockFlag(PX.RigidDynamicLockFlag.eLOCK_ANGULAR_X, !v.x);
        this.impl.setRigidDynamicLockFlag(PX.RigidDynamicLockFlag.eLOCK_ANGULAR_Y, !v.y);
        this.impl.setRigidDynamicLockFlag(PX.RigidDynamicLockFlag.eLOCK_ANGULAR_Z, !v.z);
    }

    setAllowSleep (v: boolean): void {
        if (this.isStaticOrKinematic) return;
        const st = this.impl.getSleepThreshold() as number;
        const wc = v ? Math.max(0.0, st - 0.001) : st + 0xffffffff;
        this.impl.setWakeCounter(wc);
    }

    wakeUp (): void {
        if (this.isStatic) return;
        this.impl.wakeUp();
    }

    sleep (): void {
        if (this.isStatic) return;
        this.impl.putToSleep();
    }

    clearState (): void {
        if (this.isStatic) return;
        this.clearForces();
        this.clearVelocity();
    }

    clearForces (): void {
        if (this.isStatic) return;
        this._sharedBody.clearForces();
    }

    clearVelocity (): void {
        if (this.isStatic) return;
        this._sharedBody.clearVelocity();
    }

    setSleepThreshold (v: number): void {
        if (this.isStatic) return;
        this.impl.setSleepThreshold(v);
    }

    getSleepThreshold (): number {
        if (this.isStatic) return 0;
        return this.impl.getSleepThreshold();
    }

    getLinearVelocity (out: IVec3Like): void {
        if (this.isStatic) return;
        Vec3.copy(out, this.impl.getLinearVelocity());
    }

    setLinearVelocity (value: IVec3Like): void {
        if (this.isStaticOrKinematic) return;
        this.impl.setLinearVelocity(value, true);
    }

    getAngularVelocity (out: IVec3Like): void {
        if (this.isStatic) return;
        Vec3.copy(out, this.impl.getAngularVelocity());
    }

    setAngularVelocity (value: IVec3Like): void {
        if (this.isStaticOrKinematic) return;
        this.impl.setAngularVelocity(value, true);
    }

    applyForce (force: IVec3Like, relativePoint?: IVec3Like): void {
        if (!this.isInScene || this.isStaticOrKinematic) return;
        this._sharedBody.syncSceneToPhysics();
        const rp = relativePoint || Vec3.ZERO;
        applyForce(true, this.impl, force, rp);
    }

    applyLocalForce (force: IVec3Like, relativePoint?: IVec3Like): void {
        if (!this.isInScene || this.isStaticOrKinematic) return;
        this._sharedBody.syncSceneToPhysics();
        const rp = relativePoint || Vec3.ZERO;
        applyForce(false, this.impl, force, rp);
    }

    applyImpulse (force: IVec3Like, relativePoint?: IVec3Like): void {
        if (!this.isInScene || this.isStaticOrKinematic) return;
        this._sharedBody.syncSceneToPhysics();
        const rp = relativePoint || Vec3.ZERO;
        applyImpulse(true, this.impl, force, rp);
    }

    applyLocalImpulse (force: IVec3Like, relativePoint?: IVec3Like): void {
        if (!this.isInScene || this.isStaticOrKinematic) return;
        this._sharedBody.syncSceneToPhysics();
        const rp = relativePoint || Vec3.ZERO;
        applyImpulse(false, this.impl, force, rp);
    }

    applyTorque (torque: IVec3Like): void {
        if (!this.isInScene || this.isStaticOrKinematic) return;
        applyTorqueForce(this.impl, torque);
    }

    applyLocalTorque (torque: IVec3Like): void {
        if (!this.isInScene || this.isStaticOrKinematic) return;
        this._sharedBody.syncSceneToPhysics();
        Vec3.transformQuat(v3_0, torque, this._sharedBody.node.worldRotation);
        applyTorqueForce(this.impl, v3_0);
    }

    setGroup (v: number): void {
        this._sharedBody.setGroup(v);
    }

    getGroup (): number {
        return this._sharedBody.getGroup();
    }

    addGroup (v: number): void {
        this._sharedBody.addGroup(v);
    }

    removeGroup (v: number): void {
        this._sharedBody.removeGroup(v);
    }

    setMask (v: number): void {
        this._sharedBody.setMask(v);
    }

    getMask (): number {
        return this._sharedBody.getMask();
    }

    addMask (v: number): void {
        this._sharedBody.addMask(v);
    }

    removeMask (v: number): void {
        this._sharedBody.removeMask(v);
    }
}

File: native/cocos/physics/physx/PhysXRigidBody.cpp
/****************************************************************************
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "physics/physx/PhysXRigidBody.h"
#include "physics/physx/PhysXInc.h"
#include "physics/physx/PhysXUtils.h"
#include "physics/physx/PhysXWorld.h"

using physx::PxActorFlag;
using physx::PxForceMode;
using physx::PxReal;
using physx::PxTransform;
using physx::PxVec3;

namespace cc {
namespace physics {

PhysXRigidBody::PhysXRigidBody() {
    _mObjectID = PhysXWorld::getInstance().addWrapperObject(reinterpret_cast<uintptr_t>(this));
}

void PhysXRigidBody::initialize(Node *node, ERigidBodyType t, uint32_t g) {
    _mGroup = g;
    PhysXWorld &ins = PhysXWorld::getInstance();
    _mSharedBody = ins.getSharedBody(node, this);
    getSharedBody().reference(true);
    getSharedBody().setType(t);
}

void PhysXRigidBody::onEnable() {
    _mEnabled = true;
    getSharedBody().enabled(true);
}

void PhysXRigidBody::onDisable() {
    _mEnabled = false;
    getSharedBody().enabled(false);
}

void PhysXRigidBody::onDestroy() {
    getSharedBody().reference(false);
    PhysXWorld::getInstance().removeWrapperObject(_mObjectID);
}

bool PhysXRigidBody::isAwake() {
    if (!getSharedBody().isInWorld() || getSharedBody().isStatic()) return false;
    return !getSharedBody().getImpl().rigidDynamic->isSleeping();
}

bool PhysXRigidBody::isSleepy() {
    return false;
}

bool PhysXRigidBody::isSleeping() {
    if (!getSharedBody().isInWorld() || getSharedBody().isStatic()) return true;
    return getSharedBody().getImpl().rigidDynamic->isSleeping();
}

void PhysXRigidBody::setType(ERigidBodyType v) {
    getSharedBody().setType(v);
}

void PhysXRigidBody::setMass(float v) {
    getSharedBody().setMass(v);
}

void PhysXRigidBody::setLinearDamping(float v) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setLinearDamping(v);
}

void PhysXRigidBody::setAngularDamping(float v) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setAngularDamping(v);
}

void PhysXRigidBody::useGravity(bool v) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setActorFlag(PxActorFlag::eDISABLE_GRAVITY, !v);
}

void PhysXRigidBody::useCCD(bool v) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setRigidBodyFlag(physx::PxRigidBodyFlag::eENABLE_CCD, v);
}

void PhysXRigidBody::setLinearFactor(float x, float y, float z) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setRigidDynamicLockFlag(physx::PxRigidDynamicLockFlag::eLOCK_LINEAR_X, x == 0.);
    getSharedBody().getImpl().rigidDynamic->setRigidDynamicLockFlag(physx::PxRigidDynamicLockFlag::eLOCK_LINEAR_Y, y == 0.);
    getSharedBody().getImpl().rigidDynamic->setRigidDynamicLockFlag(physx::PxRigidDynamicLockFlag::eLOCK_LINEAR_Z, z == 0.);
}

void PhysXRigidBody::setAngularFactor(float x, float y, float z) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setRigidDynamicLockFlag(physx::PxRigidDynamicLockFlag::eLOCK_ANGULAR_X, x == 0.);
    getSharedBody().getImpl().rigidDynamic->setRigidDynamicLockFlag(physx::PxRigidDynamicLockFlag::eLOCK_ANGULAR_Y, y == 0.);
    getSharedBody().getImpl().rigidDynamic->setRigidDynamicLockFlag(physx::PxRigidDynamicLockFlag::eLOCK_ANGULAR_Z, z == 0.);
}

void PhysXRigidBody::setAllowSleep(bool v) {
    if (!getSharedBody().isDynamic()) return;
    PxReal wc = v ? 0.0001F : FLT_MAX;
    getSharedBody().getImpl().rigidDynamic->setWakeCounter(wc);
}

void PhysXRigidBody::wakeUp() {
    if (!getSharedBody().isInWorld() || getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->wakeUp();
}

void PhysXRigidBody::sleep() {
    if (!getSharedBody().isInWorld() || getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->putToSleep();
}

void PhysXRigidBody::clearState() {
    if (!getSharedBody().isInWorld()) return;
    clearForces();
    clearVelocity();
}

void PhysXRigidBody::clearForces() {
    if (!getSharedBody().isInWorld()) return;
    getSharedBody().clearForces();
}

void PhysXRigidBody::clearVelocity() {
    getSharedBody().clearVelocity();
}

void PhysXRigidBody::setSleepThreshold(float v) {
    if (getSharedBody().isStatic()) return;
    //(approximated) mass-normalized kinetic energy
    float ke = 0.5F * v * v;
    getSharedBody().getImpl().rigidDynamic->setSleepThreshold(ke);
}

float PhysXRigidBody::getSleepThreshold() {
    float ke = getSharedBody().getImpl().rigidDynamic->getSleepThreshold();
    float v = sqrtf(2.F * ke);
    return v;
}

cc::Vec3 PhysXRigidBody::getLinearVelocity() {
    if (getSharedBody().isStatic()) return cc::Vec3::ZERO;
    cc::Vec3 cv;
    pxSetVec3Ext(cv, getSharedBody().getImpl().rigidDynamic->getLinearVelocity());
    return cv;
}

void PhysXRigidBody::setLinearVelocity(float x, float y, float z) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setLinearVelocity(PxVec3{x, y, z});
}

cc::Vec3 PhysXRigidBody::getAngularVelocity() {
    if (getSharedBody().isStatic()) return cc::Vec3::ZERO;
    cc::Vec3 cv;
    pxSetVec3Ext(cv, getSharedBody().getImpl().rigidDynamic->getAngularVelocity());
    return cv;
}

void PhysXRigidBody::setAngularVelocity(float x, float y, float z) {
    if (getSharedBody().isStatic()) return;
    getSharedBody().getImpl().rigidDynamic->setAngularVelocity(PxVec3{x, y, z});
}

void PhysXRigidBody::applyForce(float x, float y, float z, float rx, float ry, float rz) {
    if (!getSharedBody().isInWorld() || getSharedBody().isStaticOrKinematic()) return;
    const PxVec3 force{x, y, z};
    if (force.isZero()) return;
    auto *body = getSharedBody().getImpl().rigidDynamic;
    body->addForce(force, PxForceMode::eFORCE, true);
    const PxVec3 torque = (PxVec3{rx, ry, rz}).cross(force);
    if (!torque.isZero()) body->addTorque(torque, PxForceMode::eFORCE, true);
}

void PhysXRigidBody::applyLocalForce(float x, float y, float z, float rx, float ry, float rz) {
    if (!getSharedBody().isInWorld() || getSharedBody().isStaticOrKinematic()) return;
    const PxVec3 force{x, y, z};
    if (force.isZero()) return;
    auto *body = getSharedBody().getImpl().rigidDynamic;
    const PxTransform bodyPose = body->getGlobalPose();
    const PxVec3 worldForce = bodyPose.rotate(force);
    const PxVec3 worldPos = bodyPose.rotate(PxVec3{rx, ry, rz});
    body->addForce(worldForce, PxForceMode::eFORCE, true);
    const PxVec3 torque = worldPos.cross(worldForce);
    if (!torque.isZero()) body->addTorque(torque, PxForceMode::eFORCE, true);
}

void PhysXRigidBody::applyImpulse(float x, float y, float z, float rx, float ry, float rz) {
    if (!getSharedBody().isInWorld() || getSharedBody().isStaticOrKinematic()) return;
    const PxVec3 impulse{x, y, z};
    if (impulse.isZero()) return;
    auto *body = getSharedBody().getImpl().rigidDynamic;
    const PxVec3 torque = (PxVec3{rx, ry, rz}).cross(impulse);
    body->addForce(impulse, PxForceMode::eIMPULSE, true);
    if (!torque.isZero()) body->addTorque(torque, PxForceMode::eIMPULSE, true);
}

void PhysXRigidBody::applyLocalImpulse(float x, float y, float z, float rx, float ry, float rz) {
    if (!getSharedBody().isInWorld() || getSharedBody().isStaticOrKinematic()) return;
    const PxVec3 impulse{x, y, z};
    if (impulse.isZero()) return;
    auto *body = getSharedBody().getImpl().rigidDynamic;
    const PxTransform bodyPose = body->getGlobalPose();
    const PxVec3 worldImpulse = bodyPose.rotate(impulse);
    const PxVec3 worldPos = bodyPose.rotate(PxVec3{rx, ry, rz});
    body->addForce(worldImpulse, PxForceMode::eIMPULSE, true);
    const PxVec3 torque = worldPos.cross(worldImpulse);
    if (!torque.isZero()) body->addTorque(torque, PxForceMode::eIMPULSE, true);
}

void PhysXRigidBody::applyTorque(float x, float y, float z) {
    if (!getSharedBody().isInWorld() || getSharedBody().isStaticOrKinematic()) return;
    PxVec3 torque{x, y, z};
    if (torque.isZero()) return;
    getSharedBody().getImpl().rigidDynamic->addTorque(torque, PxForceMode::eFORCE, true);
}

void PhysXRigidBody::applyLocalTorque(float x, float y, float z) {
    if (!getSharedBody().isInWorld() || getSharedBody().isStaticOrKinematic()) return;
    PxVec3 torque{x, y, z};
    if (torque.isZero()) return;
    auto *body = getSharedBody().getImpl().rigidDynamic;
    const PxTransform bodyPose = body->getGlobalPose();
    body->addTorque(bodyPose.rotate(PxVec3{x, y, z}), PxForceMode::eFORCE, true);
}

uint32_t PhysXRigidBody::getGroup() {
    return getSharedBody().getGroup();
}

void PhysXRigidBody::setGroup(uint32_t g) {
    getSharedBody().setGroup(g);
}

uint32_t PhysXRigidBody::getMask() {
    return getSharedBody().getMask();
}

void PhysXRigidBody::setMask(uint32_t m) {
    getSharedBody().setMask(m);
}

} // namespace physics
} // namespace cc
